Mathlib/Data/Matroid/Circuit.lean

```diff
@@ -0,0 +1,173 @@
+Authors: Peter Nelson
+-/
+import Mathlib.Data.Matroid.Closure
+
+/-!
+# Matroid Circuits
+
+A `Circuit` of a matroid `M` is a minimal set `C` that is dependent in `M`.
+A matroid is determined by its set of circuits, and often the circuits
+offer a more compact description of a matroid than the collection of independent sets or bases.
+In matroids arising from graphs, circuits correspond to graphical cycles.
+
+# Main Declarations
+
+* `Matroid.Circuit M C` means that `C` is minimally dependent in `M`.
+-/
+
+variable {α : Type*} {M : Matroid α} {C C' I X Y R : Set α} {e f x y : α}
+
+open Set
+
+namespace Matroid
+
+/-- A `Circuit` of `M` is a minimal dependent set in `M` -/
+def Circuit (M : Matroid α) := Minimal M.Dep
+
+lemma circuit_def : M.Circuit C ↔ Minimal M.Dep C := Iff.rfl
+
+lemma Circuit.dep (hC : M.Circuit C) : M.Dep C :=
+  hC.prop
+
+lemma Circuit.not_indep (hC : M.Circuit C) : ¬ M.Indep C :=
+  hC.dep.not_indep
+
+lemma Circuit.minimal (hC : M.Circuit C) : Minimal M.Dep C :=
+  hC
+
+@[aesop unsafe 20% (rule_sets := [Matroid])]
+lemma Circuit.subset_ground (hC : M.Circuit C) : C ⊆ M.E :=
+  hC.dep.subset_ground
+
+lemma Circuit.nonempty (hC : M.Circuit C) : C.Nonempty :=
+  hC.dep.nonempty
+
+lemma empty_not_circuit (M : Matroid α) : ¬M.Circuit ∅ :=
+  fun h ↦ by simpa using h.nonempty
+
+lemma circuit_iff : M.Circuit C ↔ M.Dep C ∧ ∀ ⦃D⦄, M.Dep D → D ⊆ C → D = C := by
+  simp_rw [circuit_def, minimal_subset_iff, eq_comm (a := C)]
+
+lemma Circuit.ssubset_indep (hC : M.Circuit C) (hXC : X ⊂ C) : M.Indep X := by
+  rw [← not_dep_iff (hXC.subset.trans hC.subset_ground)]
+  exact fun h ↦ hXC.ne ((circuit_iff.1 hC).2 h hXC.subset)
+
+lemma Circuit.diff_singleton_indep (hC : M.Circuit C) (he : e ∈ C) : M.Indep (C \ {e}) :=
+  hC.ssubset_indep (diff_singleton_sSubset.2 he)
+
+lemma circuit_iff_forall_ssubset : M.Circuit C ↔ M.Dep C ∧ ∀ ⦃I⦄, I ⊂ C → M.Indep I := by
+  simp_rw [circuit_iff, dep_iff, and_congr_right_iff, ssubset_iff_subset_ne, and_imp]
+  exact fun _ hC ↦ ⟨fun h I hIC hne ↦ by_contra fun hi ↦ hne (h hi (hIC.trans hC) hIC),
+    fun h I hi _ hIC ↦ by_contra fun hne ↦ hi (h hIC hne)⟩
+
+lemma circuit_antichain : IsAntichain (· ⊆ ·) (setOf M.Circuit) :=
+  fun _ hC _ hC' hne hss ↦ hne <| (Circuit.minimal hC').eq_of_subset hC.dep hss
+
+lemma Circuit.eq_of_not_indep_subset (hC : M.Circuit C) (hX : ¬ M.Indep X) (hXC : X ⊆ C) :
+    X = C :=
+  eq_of_le_of_not_lt hXC (hX ∘ hC.ssubset_indep)
+
+lemma Circuit.eq_of_dep_subset (hC : M.Circuit C) (hX : M.Dep X) (hXC : X ⊆ C) : X = C :=
+  hC.eq_of_not_indep_subset hX.not_indep hXC
+
+lemma Circuit.not_ssubset (hC : M.Circuit C) (hC' : M.Circuit C') : ¬C' ⊂ C :=
+  fun h' ↦ h'.ne (hC.eq_of_dep_subset hC'.dep h'.subset)
+
+lemma Circuit.eq_of_subset_circuit (hC : M.Circuit C) (hC' : M.Circuit C') (h : C ⊆ C') : C = C' :=
+  hC'.eq_of_dep_subset hC.dep h
+
+lemma circuit_iff_dep_forall_diff_singleton_indep :
+    M.Circuit C ↔ M.Dep C ∧ ∀ e ∈ C, M.Indep (C \ {e}) := by
```

Comment:
I'd expect a similar general Minimal lemma would be useful here too

Reply:
It's the same issue; `Dep` and `¬Indep` aren't the same.

Reply:
golfed. I added a lemma about `Minimal (¬ M.Indep ·)` to make this easier.